pkg/kubelet/pleg/evented.go

```diff
@@ -125,7 +125,9 @@ func (e *EventedPLEG) Start() {
 	setEventedPLEGUsage(true)
 	e.stopCh = make(chan struct{})
 	e.stopCacheUpdateCh = make(chan struct{})
-	go wait.Until(e.watchEventsChannel, 0, e.stopCh)
+	go wait.Until(func() {
+		e.watchEventsChannel(ctx)
+	}, 0, e.stopCh)
```

Comment:
The expectation for a function taking a `ctx` is that canceling that context also cancels the function and anything started by it. This `wait.Until` doesn't do that.

The right solution would be to reconsider how `EventedPLEG` gets stopped. Perhaps `e.stopCh` can be removed?

Reply:
This should do it:
```patch
diff --git a/pkg/kubelet/pleg/evented.go b/pkg/kubelet/pleg/evented.go
index 51ffdd49b9e..d90ac2bc549 100644
--- a/pkg/kubelet/pleg/evented.go
+++ b/pkg/kubelet/pleg/evented.go
@@ -18,6 +18,7 @@ package pleg
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sync"
 	"time"
@@ -77,10 +78,8 @@ type EventedPLEG struct {
 	eventedPlegMaxStreamRetries int
 	// Indicates relisting related parameters
 	relistDuration *RelistDuration
-	// Stop the Evented PLEG by closing the channel.
-	stopCh chan struct{}
-	// Stops the periodic update of the cache global timestamp.
-	stopCacheUpdateCh chan struct{}
+	// Stop the Evented PLEG and periodic update of the cache global timestamp by canceling their context.
+	stop func(reason error)
 	// Locks the start/stop operation of the Evented PLEG.
 	runningMu sync.Mutex
 }
@@ -123,12 +122,12 @@ func (e *EventedPLEG) Start(ctx context.Context) {
 		return
 	}
 	setEventedPLEGUsage(true)
-	e.stopCh = make(chan struct{})
-	e.stopCacheUpdateCh = make(chan struct{})
-	go wait.Until(func() {
+	ctx, stop := context.WithCancelCause(ctx)
+	e.stop = stop
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
 		e.watchEventsChannel(ctx)
-	}, 0, e.stopCh)
-	go wait.Until(e.updateGlobalCache, globalCacheUpdatePeriod, e.stopCacheUpdateCh)
+	}, 0)
+	go wait.UntilWithContext(ctx, e.updateGlobalCache, globalCacheUpdatePeriod)
 }
 
 // Stop stops the Evented PLEG
@@ -139,8 +138,7 @@ func (e *EventedPLEG) Stop() {
 		return
 	}
 	setEventedPLEGUsage(false)
-	close(e.stopCh)
-	close(e.stopCacheUpdateCh)
+	e.stop(errors.New("EventedPLEG.Stop was called"))
 }
 
 // In case the Evented PLEG experiences undetectable issues in the underlying
@@ -148,7 +146,7 @@ func (e *EventedPLEG) Stop() {
 // given state while it has progressed in its life cycle. This function will be
 // called periodically to update the global timestamp of the cache so that those
 // pods stuck at GetNewerThan in pod workers will get unstuck.
-func (e *EventedPLEG) updateGlobalCache() {
+func (e *EventedPLEG) updateGlobalCache(_ context.Context) {
 	e.cache.UpdateTime(time.Now())
 }
 
```

I am not sure why the code used two separate stop channels before.